### DISCUSSION TITLE
UTXO Caching strategy predicated on power law

### DISCUSSION THREAD

**Discussion by joschisan** [utreexo/utreexo#78] [General]
From what I have read so far, the current caching strategy is predicated on the fact that the lifetimes of UTXOs follow a power law. However, this distribution seems to depend on current low transaction costs - and possibly the traffic being dominated by exchanges. The long-term security of bitcoin depends on high average fees as mining subsidies dwindle. In that case, I would expect the average transaction lifetime to increase as transaction costs rise - most of the bitcoins exchanging ownership quickly and repeatedly would do so on lightning. As most of the traffic would be driven by lightning channels and larger transactions, I would expect the distribution to transition to a normal distribution with an average lifetime of a year or longer. In that case, caching utxos by age would break down.

Furthermore, it seems that batching proofs by providing sparse subtrees cannot be used when relaying single transactions instead of blocks, as unconfirmed transactions cannot be cached. Hence we would have to provide mostly full-length inclusion proofs when relaying transactions. If most nodes add most transactions to their mempool before they receive the block, they would already have most of the proofs needed for the block in their mempool. Therefore, after IBD, most inclusion proofs are distributed in the network as additional transaction data, not as additional block data. Therefore, most proofs would be sent as unbatched, full-length proofs.

Therefore, I wondered why I did not find a discussion on the following caching strategy: For a fixed parameter n and tree of height n + k in our forest, simply cache k levels from the tree's root. Trees of height n or less are not cached. When we combine or split trees, the cache heights match up. When we share n with our peers, they can provide inclusions proofs of length at most n up to the cached tree node. By decreasing n, we could increase memory usage and decrease bandwidth usage. The performance of this strategy would neither be predicated on the utxo lifetime distribution nor on UTXOs being spent in the same block being close to each other in the accumulator. Furthermore, the higher n, the smaller the impact of batching proofs on bandwidth, and we can achieve good performance when relaying proofs as additional transaction data.

Let's do an example calculation on performance: Using 26 roots, we can accumulate about 136 million UTXOs which is sufficient for our current UTXO set. Now set n = 5. Assuming a tree node with two pointers is 32 + 8 + 8 = 48 bytes in size, this should require about 201 MB of memory as we cache up to k = 21 levels from the root. Since we now require partial proofs of length n = 5, the proof overhead for a single block would be 800KB, even if all proofs are received individually as transaction data.

Did I make an error in this line of reasoning?

**Comment by kcalvinalvin** [comment]
> From what I have read so far, the current caching strategy is predicated on the fact that the lifetimes of UTXOs follow a power law. However, this distribution seems to depend on current low transaction costs - and possibly the traffic being dominated by exchanges. The long-term security of bitcoin depends on high average fees as mining subsidies dwindle. In that case, I would expect the average transaction lifetime to increase as transaction costs rise - most of the bitcoins exchanging ownership quickly and repeatedly would do so on lightning. As most of the traffic would be driven by lightning channels and larger transactions, I would expect the distribution to transition to a normal distribution with an average lifetime of a year or longer. In that case, caching utxos by age would break down.

That could happen. But that's a big assumption.

> Furthermore, it seems that batching proofs by providing sparse subtrees cannot be used when relaying single transactions instead of blocks, as unconfirmed transactions cannot be cached. Hence we would have to provide mostly full-length inclusion proofs when relaying transactions.

Today, the current code that exists in github.com/utreexo/utreexod is doing that where the full proof is sent to the peer per tx. However, what we're planning on implementing (and later having in the spec) is having the peer ask for the proof hashes it needs. So when a utreexo node is receives a tx, it'll calculate the missing hashes and specifically ask for those hashes from the peer. This is written up in detail [here](https://github.com/utreexo/utreexo/discussions/77#discussion-4763434).

> If most nodes add most transactions to their mempool before they receive the block, they would already have most of the proofs needed for the block in their mempool. Therefore, after IBD, most inclusion proofs are distributed in the network as additional transaction data, not as additional block data. Therefore, most proofs would be sent as unbatched, full-length proofs.

That's true. The code we have now doesn't account for that but it's feasible to have something like BIP152 for utreexo proofs.

> Therefore, I wondered why I did not find a discussion on the following caching strategy: For a fixed parameter n and tree of height n + k in our forest, simply cache k levels from the tree's root. Trees of height n or less are not cached. When we combine or split trees, the cache heights match up. When we share n with our peers, they can provide inclusions proofs of length at most n up to the cached tree node. By decreasing n, we could increase memory usage and decrease bandwidth usage. The performance of this strategy would neither be predicated on the utxo lifetime distribution nor on UTXOs being spent in the same block being close to each other in the accumulator. Furthermore, the higher n, the smaller the impact of batching proofs on bandwidth, and we can achieve good performance when relaying proofs as additional transaction data.

Just to make sure I'm getting your proposal correctly.

```
28
|---------------\
24              25              26
|-------\       |-------\       |-------\
16      17      18      19      20      21      22
|---\   |---\   |---\   |---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07  08  09  10  11  12  13  14
```

For the above tree that's a height of 4, if `n == 2`, we'd cache `28`, `24`, `25` and `26`?

It's been a while since I ran an ibd test but I do think the current code ended up caching most of the higher up nodes. Making that explicit in the caching strategy isn't something we've explored and could make sense. It makes sense so I'd be curious as to how it performs against the current code.

> Let's do an example calculation on performance: Using 26 roots, we can accumulate about 136 million UTXOs which is sufficient for our current UTXO set.

26 roots as in the forest's height is 26?

Now set n = 5. Assuming a tree node with two pointers is 32 + 8 + 8 = 48 bytes in size, this should require about 201 MB of memory as we cache up to k = 21 levels from the root. Since we now require partial proofs of length n = 5, the proof overhead for a single block would be 800KB, even if all proofs are received individually as transaction data.

Could you share how you've calculated 800KB of proof per block?

**Comment by joschisan** [comment]
> That could happen. But that's a big assumption.

The exact distribution is, of course, unpredictable; however, that rising transaction costs would cause longer average UTXO lifetimes seems quite straightforward to me, does it not?  Have you investigated if there has been an effect on the distribution when transaction costs rose in the past? 

> For the above tree that's a height of 4, if n == 2, we'd cache 28, 24, 25 and 26?

This is correct.

> Could you share how you've calculated 800KB of proof per block?

I have assumed an average of 5000 inputs a block. I think I found that number somewhere in the public works on UTreeXO, but not sure where. Therefore 5000 * 5 * 32B = 800kB.

> It's been a while since I ran an ibd test but I do think the current code ended up caching most of the higher up nodes. Making that explicit in the caching strategy isn't something we've explored and could make sense. 

I actually expected the current code to cache most of the higher up nodes - this gave me the intuition for my strategy: when you choose a leaf uniformly at random from the leafs of a perfect binary tree, then the probability that a node is in the proof is 1/2 for the roots children, 1/4 for the roots grandchildren and 1/8 for the roots grand-grandchildren. Hence when we choose a couple hundred leafs the first couple of layers are included with a probability near one. Furthermore, when we choose a leaf uniformly at random from the leafs of a forest of perfect trees the probability of a leaf being in a certain tree is exponential with base two with respect to the tree's height. Since thoose probabilities need to be multiplied to derive the probability of a certain node being in the proof of a leaf choosen uniformly at random from all leafs of the forest it makes sense to cache one more level in the tree that is one level higher.
Therefore, assuming this uniform distribution, you will find that all nodes in the lowest cached level of the tree's are included in a proof with the same probability. Furthermore, those nodes in the lowest cached levels have the minimum probability of all nodes being cached and all nodes with higher or equal probability are also cached.

Let me sum up the advantages I see compared to the current strategy. 

1. Independent of distribution of UTXO Lifetimes or UTXOs spent together being close in the accumulator. Long-term Performance in terms of bandwidth/memory tradeoff is only dependent on UTXO set size. 
2.  Does not rely on batching proofs into sparse subtrees. Hence, Performance is independent of whether proofs have been received independently via transactions or blocks. 
3. This seems to me like it would simplify implementation significantly; verifying/serializing/deserializing sparse subtrees while providing only the hashes needed seems to me to be a lot more difficult than working with single proofs one at a time, which are simple lists of hashes of length less than n. I have implemented a UTreeXO accumulator supporting add, verify and delete in blockwise atomic updates while extending truncated proofs to full proofs in about 400 lines of rust, testing excluded; find the implementation here https://github.com/joschisan/utreexo. As an accumulator is consensus critical code every line creates a high cost in review and maintenance.
4. When relaying a proof, we simply extend/truncate a proof to the n requested by the peer. This reduces communication complexity. The peer's proof length limit n could be communicated when establishing a connection as I assume it would be updated rarely.
